src/stdlib/replace_with.rs

```diff
@@ -0,0 +1,214 @@
+                result: Ok("fooobar"),
+            },
+            Example {
+                title: "replace with capture group",
+                source: r#"replace_with("foo123bar", r'foo(\d+)bar') -> |m| { x = m[1]; "x={{x}}" }"#,
+                result: Ok(r#"x=123"#),
+            },
+            Example {
+                title: "process capture group",
+                source: r#"replace_with(s'Got message: {"msg": "b"}', r'message: (\{.*\})') -> |m| { to_string!(parse_json!(m[1]).msg) }"#,
+                result: Ok("Got b"),
+            },
+            Example {
+                title: "Optional capture group",
+                source: r#"replace_with("foobar", r'bar( of gold)?') -> |m| { if m[1] == null { "baz" } else { "rich" } }"#,
+                result: Ok("foobaz"),
+            },
+        ]
+    }
+
+    fn compile(
+        &self,
+        _state: &state::TypeState,
+        _ctx: &mut FunctionCompileContext,
+        arguments: ArgumentList,
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let pattern = arguments.required("pattern");
+        let count = arguments.optional("count").unwrap_or(expr!(-1));
+
+        let closure = arguments.required_closure()?;
+
+        // TODO: Should I check that the closure is not fallible?
+
+        Ok(ReplaceWithFn {
+            value,
+            pattern,
+            count,
+            closure,
+        }
+        .as_expr())
+    }
+
+    fn closure(&self) -> Option<closure::Definition> {
+        use closure::{Definition, Input, Output, Variable, VariableKind};
+
+        Some(Definition {
+            inputs: vec![Input {
+                parameter_keyword: "value",
+                kind: Kind::bytes(),
+                variables: vec![
+                    Variable {
+                        kind: VariableKind::Exact(Kind::array(Collection::from_parts(
+                                          BTreeMap::from([(0.into(), Kind::bytes())]),
+                                          Kind::bytes()))),
+                    },
+                ],
+                output: Output::Kind(Kind::bytes()),
+                example: Example {
+                    title: "replace with hash",
+                    source: r#"replace_with("received email from a@example.com", pattern: r'\w+@\w+\.\w+') -> |match| { sha2(match[0]) }"#,
+                    result: Ok("received email from 896bdca840c9304a5d0bdbeacc4ef359e3093f80c9777c9967e31ba0ff99ed58"),
+                },
+            }],
+            is_iterator: false,
+        })
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ReplaceWithFn {
+    value: Box<dyn Expression>,
+    pattern: Box<dyn Expression>,
+    count: Box<dyn Expression>,
+    closure: FunctionClosure,
+}
+
+impl FunctionExpression for ReplaceWithFn {
+    fn resolve(&self, ctx: &mut Context) -> ExpressionResult<Value> {
+        let value = self.value.resolve(ctx)?;
+        let pattern = self.pattern.resolve(ctx)?;
```

Comment:
For resolving the pattern, see this example: https://github.com/vectordotdev/vrl/pull/563/files#diff-8f4955c315c501ce024ba7c200c8af2a9aa33c59db955c305b99615690c83e9dR117-R122